pallets/liquidity-mining/src/lib.rs

```diff
@@ -0,0 +1,1325 @@
+
+				let g_pool_account = Self::pool_account_id(g_pool.id)?;
+				//NOTE: should this be 0?
+				ensure!(
+					T::MultiCurrency::free_balance(g_pool.reward_currency, &g_pool_account).is_zero(),
+					Error::<T>::RewardBalanceIsNotZero
+				);
+
+				*maybe_g_pool = None;
+
+				Self::deposit_event(Event::FarmDestroyed(farm_id, who));
+				Ok(())
+			})
+		}
+
+		#[pallet::weight(1000)]
+		pub fn withdraw_undistributed_rewards(_origin: OriginFor<T>, _farm_id: PoolId) -> DispatchResult {
+			todo!()
+		}
+
+		#[pallet::weight(1000)]
+		#[transactional]
+		pub fn add_liquidity_pool(
+			origin: OriginFor<T>,
+			farm_id: PoolId,
+			asset_pair: AssetPair,
+			multiplier: PoolMultiplier,
+			loyalty_curve: Option<LoyaltyCurve>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(!multiplier.is_zero(), Error::<T>::InvalidMultiplier);
+
+			if loyalty_curve.is_some() {
+				let c = loyalty_curve.as_ref().unwrap();
+				ensure!(
+					c.initial_reward_percentage.lt(&FixedU128::one()),
+					Error::<T>::InvalidInitialRewardPercentage
+				);
+			}
+
+			ensure!(T::AMM::exists(asset_pair), Error::<T>::AmmPoolDoesNotExist);
+
+			<GlobalPoolData<T>>::try_mutate(farm_id, |maybe_pool| -> DispatchResult {
+				let g_pool = maybe_pool.as_mut().ok_or(Error::<T>::FarmNotFound)?;
+
+				ensure!(who == g_pool.owner, Error::<T>::Forbidden);
+
+				let amm_pool_id = T::AMM::get_pair_id(asset_pair);
+				ensure!(
+					!<LiquidityPoolData<T>>::contains_key(farm_id, &amm_pool_id),
+					Error::<T>::LiquidityPoolAlreadyExists
+				);
+
+				let now_period = Self::get_now_period(g_pool.blocks_per_period)?;
+				if !g_pool.total_shares_z.is_zero() && g_pool.updated_at != now_period {
+					let reward_per_period = Self::get_global_pool_reward_per_period(
+						g_pool.yield_per_period.into(),
+						g_pool.total_shares_z,
+						g_pool.max_reward_per_period,
+					)?;
+					Self::update_global_pool(g_pool, now_period, reward_per_period)?;
+				}
+				g_pool.liq_pools_count = g_pool.liq_pools_count.checked_add(1).ok_or(Error::<T>::Overflow)?;
+
+				let liq_pool_id = Self::get_next_id()?;
+				let pallet_account = Self::account_id();
+				let (nft_class, _) = pallet_nft::Pallet::<T>::do_create_class(
+					pallet_account,
+					ClassType::LiquidityMining,
+					vec![].try_into().unwrap(),
+				)?;
+				<NftClassData<T>>::insert(nft_class, (asset_pair, 0, g_pool.id));
+
+				let pool = LiquidityPoolYieldFarm::new(liq_pool_id, now_period, loyalty_curve, multiplier, nft_class);
+
+				<LiquidityPoolData<T>>::insert(g_pool.id, &amm_pool_id, &pool);
+
+				Self::deposit_event(Event::LiquidityPoolAdded(g_pool.id, amm_pool_id, pool));
+
+				Ok(())
+			})
+		}
+
+		#[pallet::weight(1000)]
+		#[transactional]
+		pub fn update_liquidity_pool(
+			origin: OriginFor<T>,
+			farm_id: PoolId,
+			asset_pair: AssetPair,
+			multiplier: PoolMultiplier,
+		) -> DispatchResult {
+			//TODO: this fn is WIP, it's not tested
```

Comment:
TODO?

Reply:
This was TODO. It's finished now and ready for review